pkg/virt-launcher/virtwrap/network/network.go

```diff
@@ -26,112 +26,79 @@ package network
 */
 
 import (
-	"fmt"
-
 	v1 "kubevirt.io/client-go/api/v1"
 	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
 )
 
 const primaryPodInterfaceName = "eth0"
 
 var vifCacheFile = "/proc/%s/root/var/run/kubevirt-private/vif-cache-%s.json"
-var podNICFactory = newpodNIC
 
 type PodCacheInterface struct {
 	Iface  *v1.Interface `json:"iface,omitempty"`
 	PodIP  string        `json:"podIP,omitempty"`
 	PodIPs []string      `json:"podIPs,omitempty"`
 }
 
-// Network configuration is split into two parts, or phases, each executed in a
-// different context.
-// Phase1 is run by virt-handler and heavylifts most configuration steps. It
-// also creates the tap device that will be passed by name to virt-launcher,
-// thus allowing unprivileged libvirt to consume a pre-configured device.
-// Phase2 is run by virt-launcher in the pod context and completes steps left
-// out of virt-handler. The reason to have a separate phase for virt-launcher
-// and not just have all the work done by virt-handler is because there is no
-// ready solution for DHCP server startup in virt-handler context yet. This is
-// a temporary limitation and the split is expected to go once the final gap is
-// closed.
-// Moving all configuration steps into virt-handler will also allow to
-// downgrade privileges for virt-launcher, specifically, to remove NET_ADMIN
-// capability. Future patches should address that. See:
-// https://github.com/kubevirt/kubevirt/issues/3085
-type podNIC interface {
-	PlugPhase1(vmi *v1.VirtualMachineInstance, iface *v1.Interface, network *v1.Network, podInterfaceName string, pid int) error
-	PlugPhase2(vmi *v1.VirtualMachineInstance, iface *v1.Interface, network *v1.Network, domain *api.Domain, podInterfaceName string) error
+type VMNetworking struct {
+	vmi     *v1.VirtualMachineInstance
+	handler NetworkHandler
 }
 
-func getNetworksAndCniNetworks(vmi *v1.VirtualMachineInstance) (map[string]*v1.Network, map[string]int) {
-	networks := map[string]*v1.Network{}
-	cniNetworks := map[string]int{}
-	for _, network := range vmi.Spec.Networks {
-		networks[network.Name] = network.DeepCopy()
-		if networks[network.Name].Multus != nil && !networks[network.Name].Multus.Default {
-			// multus pod interfaces start from 1
-			cniNetworks[network.Name] = len(cniNetworks) + 1
-		}
+func NewVMNetworking(vmi *v1.VirtualMachineInstance) *VMNetworking {
```

Comment:
I think it is preferable to have the caller inject the handler via the constructor. While I understand you can later on
```golang
# unit test code
networkConfigurator := NewVMNetworking(vmi)
networkConfigurator.handler = FakeNetworkHandler{}
```

in the unit tests, I find it a bit uglier. Won't push for it though.

Reply:
we have have a pair of constructors `NewVMNetworking` and `NewVMNetworkWIthHandler` I will do that.